codeworld-api/src/CodeWorld/Driver.hs

```diff
@@ -378,13 +404,39 @@ findTopPicture ctx ds pic = case pic of
             else return Nothing
     where map2 = fmap . fmap
 
+isDebugModeActive :: IO Bool
+isDebugModeActive = js_isDebugModeActive
+
+setDebugModeActive :: Bool -> IO ()
+setDebugModeActive True  = js_startDebugMode
+setDebugModeActive False = js_stopDebugMode
+
+sweetAlert :: Text -> Text -> Text -> IO ()
+sweetAlert style title msg = js_sweetAlert (textToJSString style) (textToJSString title)
+                                           (textToJSString msg)
+
 -- Canvas.isPointInPath does not provide a way to get the return value
 -- https://github.com/ghcjs/ghcjs-base/blob/master/JavaScript/Web/Canvas.hs#L212
 foreign import javascript unsafe "$3.isPointInPath($1,$2)"
     js_isPointInPath :: Double -> Double -> Canvas.Context -> IO Bool
 
-foreign import javascript unsafe "initDebugMode($1)"
-    js_initDebugMode :: Callback (JSVal -> JSVal -> IO ()) -> IO ()
+foreign import javascript unsafe "$3.isPointInStroke($1,$2)"
+    js_isPointInStroke :: Double -> Double -> Canvas.Context -> IO Bool
+
+foreign import javascript unsafe "initDebugMode($1,$2)"
+    js_initDebugMode :: Callback (JSVal -> IO JSVal) -> Callback (JSVal -> IO ()) -> IO ()
+
+foreign import javascript unsafe "window.debugMode"
+    js_isDebugModeActive :: IO Bool
+
+foreign import javascript unsafe "startDebugMode()"
+    js_startDebugMode :: IO ()
+
+foreign import javascript unsafe "stopDebugMode()"
+    js_stopDebugMode :: IO ()
+
+foreign import javascript unsafe "parent.sweetAlert($2,$3,$1);"
```

Comment:
I'd like to move the API boundary a bit further back.  The fact that we're using SweetAlert definitely is a detail of the environment, and shouldn't be fixed that way in external code.  So please define a function in the parent frame, and call that, rather than generating JavaScript that assumes SweetAlert is loaded.